### DISCUSSION TITLE
Codespaces keyboard mapping seems not to work in terminal

### DISCUSSION THREAD

**Discussion by dsyer** [community/community#5997] [Codespaces]
Editors work OK but the terminal thinks I have a US keyboard layout and there seems to be no way to change it.

**Comment by bamurtaugh** [comment]
Thanks @dsyer. Does it look like this could be https://github.com/microsoft/vscode/issues/24166? Also, are you working on web or desktop codespaces?

**Comment by dsyer** [comment]
I'm on web (the "Open in VSCode" button doesn't work for me, but that's a separate issue). I don't know if that's the same problem, and it's hard to know how to tell. I didn't get the impression that there was a reliable workaround either - is that wrong? Anyway, setting the keyboard dispatch to "keyCode" didn't fix it for me. I don't know how to ask the terminal what locale it thinks it is in (there's no `setxkbmap` in the container). How can I diagnose?

**Comment by dsyer** [comment]
I guess X11 keyboard utilities don't make sense anyway in the container:

```
$ sudo apt-get install x11-xkb-utils
...
$ setxkbmap -query
Cannot open display "default display"
$ setxkbmap -layout gb
Cannot open display "default display"
```

So I don't know how to fix it.

**Comment by bamurtaugh** [comment]
@tyriar @meganrogge thoughts?

**Comment by meganrogge** [comment]
@dsyer not sure if this is what you're looking for, but you can run the below to see what locale is being used 

```
echo $LANG
```

and then to set it, you can run 

```
 export LANG=ru_RU.KOI8-R (whichever locale you want)
```

**Reply by dsyer** [reply]
I tried that already. No change. Probably the terminal has to be reset somehow too, but I didn’t find any way to do that. Most of the advice from the internet assumes you have systemd, and these containers do not.

**Reply by Tyriar** [reply]
Try setting `"terminal.integrated.detectLocale": "off"`

**Reply by dsyer** [reply]
If I do that then `LANG` is empty in a new terminal, but the keyboard layout is still wrong.

**Reply by Tyriar** [reply]
@dsyer it's typically up to your machine to set LANG properly, `terminal.integrated.detectLocale` sets it for you if it looks wrong but there's only so much we can do to fill in the correct language. If setting it manually though didn't fix it I don't know what the issue is.

What locale are you expecting and why are you expecting it?

**Reply by dsyer** [reply]
I guess I'm expecting `en_GB` (because my keyboard is UK). Remember the editor windows in codespaces work just fine, so the keyboard and the browser, or whatever it is that controls the metadata, is sending the right signals, they just aren't getting used correctly by the terminal. Also `terminal.integrated.detectLocale` is setting `LANG=en_US.UTF-8`, so it's even getting that wrong, not that changing it seems to help.

**Reply by Tyriar** [reply]
@rebornix I think you looked into this before, do you know if it's possible via browser APIs to get the keyboard locale?

**Reply by rebornix** [reply]
@Tyriar there isn't any way to check the locale but we can check against `keyboard.getLayoutMap()` and guess.

**Reply by Tyriar** [reply]
Hmm, I think we'll keep it as you need to set LANG yourself in this sort of case for now.

**Reply by Tyriar** [reply]
Created https://github.com/microsoft/vscode/issues/135960 to track
